hxmath/math/DualQuaternion.hx

```diff
@@ -0,0 +1,622 @@
+package hxmath.math;
+// initial version based on nanjizal/geom/matrix/DualQuaternion
+// @author: Nanjizal
+typedef DualQuaternionShape = 
+{
+    public var real:Quaternion;
+    public var dual:Quaternion;
+}
+
+/**
+ * The default underlying type.
+ */
+class DualQuaternionDefault
+{
+    public var real:Quaternion;
+    public var dual:Quaternion;
+    
+    public function new(real:Quaternion, dual:Quaternion)
+    {
+        this.real = real;
+        this.dual = dual;
+    }
+    
+    public function toString():String
+    {
+        return '[$real, $dual]';
+    }
+}
+
+typedef DualQuaternionType = DualQuaternionDefault;
+
+/**
+ * DualQuaternion for 3D transformation, similar to a Vector4 and Quaternion
+ */
+@:forward(real,dual)
+abstract DualQuaternion(DualQuaternionType) from DualQuaternionType to DualQuaternionType
+{
+    // The number of elements in this structure ( 4 real + 4 dual )
+    public static inline var elementCount:Int = 8;
+    
+    // Zero dual quaternion 
+    public static var zero(get, never):DualQuaternion;
+    
+    // One/identity dual quaternion 
+    public static var identity(get, never):DualQuaternion;
+    
+    // Gets the corresponding transform matrix
+    public var matrix(get, never):Matrix4x4;
+    
+    // Magnitude
+    public var length(get, never):Float;
+    
+    // Dual Quaternion dotted with itself
+    public var lengthSq(get, never):Float;
+    
+    // The normalized dual quaternion
+    public var normal(get, never):DualQuaternion;
+    
+    /**
+     * Constructor.
+     * 
+     * @param real  Quaternion Real
+     * @param dual  Quaternion Dual
+     */
+    public inline function new(real:Quaternion, dual:Quaternion) 
+    {
+        this = new DualQuaternionDefault(real, dual);
+    }
+    
+    /**
+     * Construct a DualQuaternion the easy way
+     * from a rotational Quaternion and a translation Vector4
+     * 
+     * @param r         The Real Quaternion ( angle )
+     * @param t         The Vector4 ( translation )
+     * @return          The DualQuaternion
+     */
+    public static inline function create(r: Quaternion, t: Vector4 ):DualQuaternion
+    {
+        var real = r.normalize();
+        var q = new Quaternion( 0., t.x, t.y, t.z );
+        q = q*0.5;
+        var dual = q * real;
+        return new DualQuaternion(real, dual);
+    }
+    
+    /**
+     * Construct a DualQuaternion using axis angle 
+     * axis angle in degrees and axis vector3, translation in vector4
+     * 
+     * @param angleDegrees  The angle to rotate in degrees.
+     * @param axis          The axis to rotate around.
+     * @param t             The Vector4 ( translation )
+     * @return              The DualQuaternion
+     */
+    public static inline function fromAxisAngle( angleDegrees:Float
```

Comment:
Minor: should match existing function style (in this case just put it all on the same line)

Spacing below should also match existing.